### PR TITLE
[INV-22] Add tax and discount on line items field.

### PR DESCRIPTION
**Pull request by Azher9767**
add tax and discount on line items fields work in progress.

### CODE REVIEW

File: app/controllers/invoices/calculations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      
+      tax_and_discount_polys = calculation_params[:taxAndDiscountPolyAttributes].map do |td_poly|
+        TaxAndDiscountPoly.new(
+          amount: td_poly[:amount].to_f,
+          td_type: td_poly[:td_type],
+          name: td_poly[:name]
+        )
+      end 
+      
+      @total, @sub_total = InvoiceAmountCalculator.new(line_items, tax_and_discount_polys).call
```
Comment (ahmad19) [inline app/controllers/invoices/calculations_controller.rb:34]: The create and update are exactly the same methods. Lets remove the update one.

File: app/controllers/invoices/line_items_controller.rb [hunk 2]
```diff
@@ -9,6 +9,7 @@ def create
           unit_rate: @product.unit_rate,
           product_id: @product.id,
         )
+        @line_item.tax_and_discount_polies.build
```
Comment (ahmad19) [inline app/controllers/invoices/line_items_controller.rb:12]: why is this ?

Comment (ahmad19) [reply app/controllers/invoices/line_items_controller.rb:12]: If this is to display TD dropdown with line items, consider to test the implementation again as i worked on handling it in the view few days ago.

File: app/controllers/invoices_controller.rb [hunk 3]
```diff
@@ -61,19 +70,36 @@ def destroy
...
+  # Only allow a list of trusted parameters through.
+  def invoice_params
+    params[:invoice][:line_items_attributes].each do |_key, value|
+      value[:tax_and_discount_polies_attributes] = value[:tax_and_discount_polies_attributes].each do |_k, v|
+        v[:tax_and_discount_id] = v[:tax_and_discount_id].reject!(&:empty?)
+      end
+    end if params[:invoice][:line_items_attributes].present?
+    puts "=========After clean up==========="
+    puts params[:invoice]
+    puts "=================================="
```
Comment (ahmad19) [inline app/controllers/invoices_controller.rb:92]: Get rid of this debug statements.

File: app/models/line_item.rb [hunk 4]
```diff
@@ -4,7 +4,19 @@ class LineItem < ApplicationRecord
...
     quantity * unit_rate
   end
+  
+
+  private
+
+  def total_tax_percentage
+    tax_policies = tax_and_discount_polies.select { |td| td.td_type == 'tax' }
+    tax_policies.sum { |td| td.amount }
+  end
```
Comment (ahmad19) [inline app/models/line_item.rb:21]: remove it.

File: app/models/tax_and_discount_poly.rb [hunk 5]
```diff
@@ -1,5 +1,6 @@
 class TaxAndDiscountPoly < ApplicationRecord
+  belongs_to :tax_and_discount
   belongs_to :tax_discountable, polymorphic: true
 
- include TaxAndDiscountQueries
+  include TaxAndDiscountQueries
```
Comment (ahmad19) [inline app/models/tax_and_discount_poly.rb:5]: why extra space here?

File: app/services/checkers.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  def any_line_item_with_tds?
+    @line_items.any? { |line_item| line_item.tax_and_discount_polies.present? }
+  end
+
+  alias_method :td_in_line_items_only?, :any_line_item_with_tds?
+
+  def td_on_invoice_only?
+    @itds.any? && !any_line_item_with_tds?
+  end
+end
```
Comment (ahmad19) [inline app/services/checkers.rb:19]: can you double check if all these methods are being used ?

File: app/services/invoice_amount_calculator.rb [hunk 7]
```diff
@@ -1,23 +1,36 @@
...
+  #   @line_items.select do |line_item|
+  #     line_item.tax_and_discount_polies.blank? || line_item.tax_and_discount_polies.any? {|li| li.discount? }
+  #   end
+  # end
+
+  # def td_line_items
+  #   @line_items.select do |line_item|
+  #     line_item.tax_and_discount_polies.present? && line_item.tax_and_discount_polies.any? {|li| li.tax? || li.discount? }
+  #   end
+  # end
```
Comment (ahmad19) [inline app/services/invoice_amount_calculator.rb:35]: Remove

File: config/routes.rb [hunk 8]
```diff
@@ -8,7 +8,7 @@
   namespace :invoices do
     resources :line_items, only: [:create, :destroy]
     resources :tax_and_discount_polies, only: [:create, :destroy]
-    resources :calculator, only: [:create]
+    resource :calculations, only: [:create, :update]
```
Comment (ahmad19) [inline config/routes.rb:11]: remove it

**Review by ahmad19 — CHANGES_REQUESTED** [review]
Added few comments, have a look

File: app/services/checkers.rb (hunk 6 above)
Comment (Azher9767) [reply app/services/checkers.rb:19]: yes all the methods are being used I verified it.

**Pull request merged**